src/include/kdbprivate.h

```diff
@@ -210,6 +128,32 @@ struct _Key
 	 */
 	size_t dataSize;
 
+	/**
+	 * Reference counter
+	 */
+	uint16_t refs;
+
+	/**
+	 * Is this structure and its data stored in an mmap()ed memory area?
```

Comment:
So if a key data is changed you malloc a completely new keydata struct and then you malloc the data * afterwards, instead of just adjusting the values in the keydata struct and doing one malloc?

Reply:
Yes, if the flag `isInMmap` is set. You can see this logic in the `*Detach*` functions.

Reply:
I guess this can be done ... but why not reuse the struct in the mmap region?
Are there any benefits, except that we have one `if`-statement less?

Reply:
No, it just never crossed my mind that this could be done.

Reply:
> Are there any benefits, except that we have one if-statement less?

I think we could do something like this later on (pseudo-code!):

```c
void keySetData (Key * key, void * data, size_t dataSize) {
    KeyData * keyData = malloc (sizeof(KeyData) + dataSize);
    keyData->data = keyData + 1; // pointer arithmetic in terms of sizeof(KeyData)
    keyData->dataSize = dataSize;
    memcpy (keyData->data, data, dataSize);
    key->keyData = keyData;
}
```

Using only one allocation could improve cache locality (especially for small `dataSize`). Although that would need benchmarking. In any case the option is there. If `keyData->data` can be malloc when `keyData` is mmap this would be harder to do.

Reply:
> Using only one allocation could improve cache locality

Good idea, but wouldn't there be alignment problems? Isn't char * byte-aligned while the struct would be aligned as the most strict member is aligned, aka pointers, thus 4 or 8 bytes in most cases?

Reply:
> Good idea, but wouldn't there be alignment problems?

Well first of all the alignment problems only come up, when you do stuff on the stack. `malloc` always uses the biggest possible alignment, because it doesn't know what alignment you want. So the `keyData` pointer is definitely aligned correctly.

Technically, there could be a problem here, because I used `void * data`. But with `char * data` it would definitely be okay, because `keyData->data` can't be misaligned for a `char *`. If we want to support bigger alignments, we'd have to insert a padding. I'm not gonna post the code here, but [here](https://git.uclibc.org/uClibc/tree/libc/stdlib/malloc/malloc.c) is a simple implementation of malloc. From there we could extract the part the ensures the alignment (just some macros that round the size to the correct alignment).

Reply:
You're right, I might have missed it alltogether: if the struct comes first and is strictly aligned due to the pointer members, anything that comes after it will have the same nice alignment.

Sorry for the confusion. I had this mistaken, as in the mmap regions usually something else comes after the data again, which could then be misaligned due to `char *`. This is not the case here.

Were you planning to work on this, i.e. would you mind if I do some benchmarks regarding this approach?

Reply:
No, no plans here, just an idea I had. Feel free to work on this. I'd love to see benchmarks, to know whether this actually has any benefit.